lxd/storage/drivers/generic_vfs.go

```diff
@@ -304,10 +304,16 @@ func genericVFSCreateVolumeFromMigration(d Driver, initVolume func(vol Volume) (
 		return rsync.Recv(path, conn, wrapper, volTargetArgs.MigrationType.Features)
 	}
 
-	recvBlockVol := func(volName string, conn io.ReadWriteCloser, path string) error {
+	recvBlockVol := func(vol Volume, conn io.ReadWriteCloser, path string) error {
+		// Only receive block if it doesn't already exist. When using the lxd-migrate tool,
+		// the block volume is received in recvFSVol, making this function obsolete.
+		if vol.volType == VolumeTypeVM && shared.PathExists(path) {
+			return nil
+		}
```

Comment:
How does that work? Do you mean we just send one chunk through rsync which includes the block entry as a file?

If so, then this can't work against ZFS, LVM or Ceph which all have the block stored completely separately.

Reply:
> How does that work? Do you mean we just send one chunk through rsync which includes the block entry as a file?

Yes, exactly. I tried doing it the exact way LXD does migration, but it gets stuck at receiving the filesystem. Therefore, I went with the approach.

> If so, then this can't work against ZFS, LVM or Ceph which all have the block stored completely separately.

How should we go about this then? In the case of a VM, `lxd-migrate` will only want to send the image. LXD's normal migration expects a filesystem (backup.yaml, etc.) before wanting to receive the block.

Reply:
Send an empty rsync for the filesystem and then send the block.

Reply:
> Send an empty rsync for the filesystem and then send the block.

That's what I tried, and it didn't seem to work.

Reply:
Hmm, then we'll need to try harder as that should definitely work.

When LXD does the migration itself it sends:
 - filesystem through rsync (usually just includes backup.yaml and some agent files)
 - sends the block as a second step using the dd migration logic

So I'd expect us to do the same with lxd-migrate

Reply:
Now, `lxd-migrate` does it correctly (first empty rsync, then block volume). At the end however the websocket is closed for some reason, not finishing the migration. Could you take a quick look? I don't understand what the problem could be.

Reply:
I've been comparing the logic to `lxd/storage/drivers/generic_vfs.go` which has a `genericVFSMigrateVolume` that does roughly the same thing (well, bit more complex because of snapshots, bwlimit and progress tracking).

The main difference I'm seeing is the `conn.Close()` after the block has been copied.

Reply:
If we're not getting that far in the logic, then maybe take a close look at the rsync logic to see if there isn't some difference there which may cause the connection to be closed before we make it to the block part?